### ISSUE TITLE
Cache JavaScript DOM selections

### ISSUE THREAD

**Issue by danielesalvatore** [federicorosati1994/calculator_material-design#11]
In order to optimize the JavaScript execution, cache the DOM selection as following:

<pre><code>
var display = document.getElementById("display");

...

// funzione che riporta numeri e operatori nell'espressione e nel display
function n(number) {
  //display.dir = "rtl"; //o .style.direction = "rtl";
  // se il numero è 0 o si è gia ottenuto il risultato dell'operazione inizia da capo
  if(display.value =="0" || display.value==null || setup == true){
    display.value = number;
  
...

</code></pre>

Can you explain why this is an optimization?

For reference:
http://www.phpied.com/dom-access-optimization/

**Comment by danielesalvatore** [comment]
@FrancescoEsposito Help us solve this!

**Comment by federicorosati1994** [comment]
Yes, this could be obviously an optimization because with the var display you don't have to write every time "document.getElementById("display")", indeed we did this optimization! But, we didn't know how it was possible, in this way, the function n() didn't work... We'll try another time and let you know!

**Comment by danielesalvatore** [comment]
Ok I see your point but it is not exactly what I meant. Check [this JavaScript performance test](https://jsperf.com/document-getelementbyid-vs-jquery/3) in which you can benchmark and compare how many times per second you can perform a certain operation in JS. Check how many times per second you browser can manipulate the DOM and find a specific element out of it by its id. With the browse I'm currently using the result is almost 14.000 per sec. Quite a pretty result, but i wonder: even if this sounds like something quite fast, if we do it only once (cache the reference of the specific DOM element within a variable) wouldn't it be even a better situation?
What I mean is that we should limit the DOM interaction as much as possible.

PS: We will check together why your code breaks if you use a variable to cache the display reference.

**Comment by federicorosati1994** [comment]
Auguri prof! :D

**Comment by danielesalvatore** [comment]
01000111 01110010 01100001 01111010 01101001 01100101 00100000 01101101 01101001 01101100 01101100 01100101 00101100 00100000 01000110 01100101 01100100 01100101 00100001 00100000 01100011 01101001 00100000 01110110 01100101 01100100 01101001 01100001 01101101 01101111 00100000 01100100 01101111 01110000 01101111 00100000 01100001 00100000 01101100 01100101 01111010 01101001 01101111 01101110 01100101 00100001
